programs/mmm/src/instructions/ext_vanilla/ext_fulfill_buy.rs

```diff
@@ -0,0 +1,356 @@
+    pub token_program: Interface<'info, TokenInterface>,
+    pub associated_token_program: Program<'info, AssociatedToken>,
+    pub rent: Sysvar<'info, Rent>,
+    // Remaining accounts
+    // Branch: using shared escrow accounts
+    //   0: m2_program
+    //   1: shared_escrow_account
+    //   2+: creator accounts
+    // Branch: not using shared escrow accounts
+    //   0+: creator accounts
+}
+
+pub fn handler<'info>(
+    ctx: Context<'_, '_, '_, 'info, ExtSolFulfillBuy<'info>>,
+    args: SolFulfillBuyArgs,
+) -> Result<()> {
+    let token_program = &ctx.accounts.token_program;
+    let system_program = &ctx.accounts.system_program;
+    let associated_token_program = &ctx.accounts.associated_token_program;
+    let rent = &ctx.accounts.rent;
+    let pool = &mut ctx.accounts.pool;
+    let sell_state = &mut ctx.accounts.sell_state;
+    let owner = &ctx.accounts.owner;
+    let referral = &ctx.accounts.referral;
+    let payer = &ctx.accounts.payer;
+    let payer_asset_account = &ctx.accounts.payer_asset_account;
+    let asset_mint = &ctx.accounts.asset_mint;
+    let buyside_sol_escrow_account = &ctx.accounts.buyside_sol_escrow_account;
+    let pool_key = pool.key();
+    let buyside_sol_escrow_account_seeds: &[&[&[u8]]] = &[&[
+        BUYSIDE_SOL_ESCROW_ACCOUNT_PREFIX.as_bytes(),
+        pool_key.as_ref(),
+        &[ctx.bumps.buyside_sol_escrow_account],
+    ]];
+    let remaining_accounts = ctx.remaining_accounts;
+
+    check_allowlists_for_mint_ext(
+        &pool.allowlists,
+        &asset_mint.to_account_info(),
+        args.allowlist_aux,
+    )?;
+
+    let (total_price, next_price) =
+        get_sol_total_price_and_next_price(pool, args.asset_amount, true)?;
+    let seller_receives = {
+        let lp_fee_bp = get_lp_fee_bp(pool, buyside_sol_escrow_account.lamports());
+        get_buyside_seller_receives(
+            total_price,
+            lp_fee_bp,
+            0, // metadata_royalty_bp
+            0, // buyside_creator_royalty_bp,
+        )
+    }?;
+    let lp_fee = get_sol_lp_fee(pool, buyside_sol_escrow_account.lamports(), seller_receives)?;
+
+    assert_valid_fees_bp(args.maker_fee_bp, args.taker_fee_bp)?;
+    let maker_fee = get_sol_fee(seller_receives, args.maker_fee_bp)?;
+    let taker_fee = get_sol_fee(seller_receives, args.taker_fee_bp)?;
+    let referral_fee = u64::try_from(
+        maker_fee
+            .checked_add(taker_fee)
+            .ok_or(MMMErrorCode::NumericOverflow)?,
+    )
+    .map_err(|_| MMMErrorCode::NumericOverflow)?;
+
+    // withdraw sol from M2 first if shared escrow is enabled
+    if pool.using_shared_escrow() {
+        check_remaining_accounts_for_m2(remaining_accounts, &pool.owner.key())?;
+
+        let amount: u64 = (total_price as i64 + maker_fee) as u64;
+        withdraw_m2(
+            pool,
+            ctx.bumps.pool,
+            buyside_sol_escrow_account,
+            index_ra!(remaining_accounts, 1),
```

Comment:
if we are using shared escrow, we cannot directly pass in `remaining_accounts` to `invoke_transfer_checked` helper because it will contain not only the transfer hook accounts but also the m2 accounts

Reply:
make sense!